rdr_service/services/consent/validation.py

```diff
@@ -13,6 +15,171 @@
+        )
+        for result in previous_results:
+            organized_previous_results[result.participant_id][result.type].append(result)
+
+        results_to_update = []
+        for participant_id, consent_type_dict in self.results.items():
+            for consent_type, result_list in consent_type_dict.items():
+                ready_for_sync = self._find_file_ready_for_sync(result_list)
+                previous_type_list: Collection[ParsingResult] = organized_previous_results[participant_id][consent_type]
+                if ready_for_sync:
+                    for result in previous_type_list:
+                        if result.sync_status == ConsentSyncStatus.NEEDS_CORRECTING:
+                            result.sync_status = ConsentSyncStatus.OBSOLETE
+                            results_to_update.append(result)
+                    results_to_update.append(ready_for_sync)
+                else:
+                    for possible_new_result in result_list:
+                        if not any([possible_new_result.file_path == previous_result.file_path
+                                    for previous_result in previous_type_list]):
+                            results_to_update.append(possible_new_result)
+
+        self.consent_dao.batch_update_consent_files(self.session, results_to_update)
+
+    @classmethod
+    def _find_file_ready_for_sync(cls, results: List[ParsingResult]):
+        for result in results:
+            if result.sync_status == ConsentSyncStatus.READY_FOR_SYNC:
+                return result
+
+        return None
+
+
+class LogResultStrategy(ValidationOutputStrategy):
+    def __init__(self, logger, verbose, storage_provider: GoogleCloudStorageProvider):
+        self.logger = logger
+        self.verbose = verbose
+        self.storage_provider = storage_provider
+        self.results = self._build_consent_list_structure()
+
+    def add_result(self, result: ParsingResult):
+        self.results[result.participant_id][result.type].append(result)
+
+    def process_results(self):
+        report_lines = []
+        for validation_categories in self.results.values():
+            if self.verbose:
+                report_lines.append('')
+            for result_list in validation_categories.values():
+                for result in result_list:
+                    report_lines.append(self._line_output_for_validation(result, verbose=self.verbose))
+
+        self.logger.info('\n'.join(report_lines))
+
+    def _line_output_for_validation(self, file: ParsingResult, verbose: bool):
+        output_line = StringIO()
+        if verbose:
+            output_line.write(f'{str(file.id).ljust(8)} - ')
+        output_line.write(f'P{file.participant_id} - {str(file.type).ljust(10)} ')
+
+        if not file.file_exists:
+            output_line.write('missing file')
+        else:
+            errors_with_file = []
+            if not file.is_signature_valid:
+                errors_with_file.append('invalid signature')
```

Comment:
For future consideration:
I think we should be defining a dict object with all the specific errors we know how to detect so they are centralized in one place

Reply:
Yeah. I also liked some of the ideas we were discussing about not having these as hardcoded strings. We'll definitely want to revisit this.